### PR TITLE
Add eurobert

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?

### CODE REVIEW

**Comment by tomaarsen** [comment]
@ArthurZucker What's the status here? Would love to get this fully out, also so we can add ONNX support for EuroBERT: https://github.com/huggingface/optimum/pull/2321

- Tom Aarsen

**Comment by ArthurZucker** [comment]
Got caught up in the release I just needed to add integration tests making sure it works well!

File: tests/models/eurobert/test_modeling_eurobert.py [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+        with self.assertRaises(KeyError):
+            config = _reinitialize_config(base_config, {"rope_scaling": {"rope_type": "linear"}})  # missing "factor"
+
+
+@require_torch_accelerator
+class EuroBertIntegrationTest(unittest.TestCase):
+    # This variable is used to determine which CUDA device are we using for our runners (A10 or T4)
+    # Depending on the hardware we get different logits / generations
+    cuda_compute_capability_major_version = None
+
```
Comment (ArthurZucker) [inline tests/models/eurobert/test_modeling_eurobert.py:583]:
```python
from transformers import AutoTokenizer, AutoModelForMaskedLM

model_id = "EuroBERT/EuroBERT-2.1B"

tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForMaskedLM.from_pretrained(model_id, trust_remote_code=True)

text = "The capital of France is <|mask|>."
inputs = tokenizer(text, return_tensors="pt")
outputs = model(**inputs)

# To get predictions for the mask:
masked_index = inputs["input_ids"][0].tolist().index(tokenizer.mask_token_id)
predicted_token_id = outputs.logits[0, masked_index].argmax(axis=-1)
predicted_token = tokenizer.decode(predicted_token_id)
print("Predicted token:", predicted_token)
# Predicted token:  Paris

```
will run this with trust remote code

**Comment by tomaarsen** [comment]
@ArthurZucker 🏴‍☠️🏴‍☠️🏴‍☠️ I'm taking over this PR! 🏴‍☠️🏴‍☠️🏴‍☠️

Earlier today, Jina released a line of embedding models using EuroBERT:
* https://huggingface.co/jinaai/jina-embeddings-v5-text-nano-retrieval
* https://huggingface.co/jinaai/jina-embeddings-v5-text-nano-clustering
* https://huggingface.co/jinaai/jina-embeddings-v5-text-nano-text-matching
* https://huggingface.co/jinaai/jina-embeddings-v5-text-nano-classification

And it became very apparent that this PR should really be pushed over the finish line (there's [378 eurobert models](https://huggingface.co/models?other=eurobert)!). I updated the implementation considerably, mostly simplifying using `create_bidirectional_mask` & `is_causal=False` on the `config` (possible thanks to #43705). This meant I could totally defer to Llama for the `EuroBertAttention` `forward`. `EuroBertForMaskedLM` needed a tad more work, it was using `LlamaForCausalLM` via modular, but that was never a great choice. I also renamed `clf_pooling` to `classifier_pooling`, matching the config name and ModernBERT.

Beyond that, it was mostly modernization by updating e.g. the `**kwargs`, `@auto_docstring`, `rope` updates, etc.

I also updated the tests, dropping CausalLM tests (apart from some RoPE ones that are still interesting), and adding Integration tests for the 4 classes, with these models:
* EuroBertForMaskedLM: https://huggingface.co/EuroBERT/EuroBERT-210m (expected logits are based on the `trust_remote_code` implementation)
* EuroBertModel: https://huggingface.co/EuroBERT/EuroBERT-210m (expected logits are based on the `trust_remote_code` implementation)
* EuroBertForSequenceClassification: https://huggingface.co/hf-internal-testing/tiny-random-EuroBertForSequenceClassification
* EuroBertForTokenClassification: https://huggingface.co/hf-internal-testing/tiny-random-EuroBertForTokenClassification

Let's get this finished!

P.s.: The original implementation stopped working due to upgrades to our rope implementation: https://huggingface.co/EuroBERT/EuroBERT-210m/discussions/22
And even more parts will be deprecated in the future as the original implementation uses the `AttentionMaskConverter`: https://huggingface.co/EuroBERT/EuroBERT-210m/blob/main/modeling_eurobert.py#L486

cc @Cyrilvallez if you have some bandwidth, then please help me review!

- Tom Aarsen

File: tests/models/eurobert/test_modeling_eurobert.py (hunk 1 above)
Comment (tomaarsen) [reply tests/models/eurobert/test_modeling_eurobert.py:583]: Done in 9d718ada5a53b43825ff39c51e277c865cbde4f0

**Comment by tomaarsen** [comment]
Would love a final review on this!

- Tom Aarsen

**Comment by contrebande-labs** [comment]
Yes please!

**Comment by ArthurZucker** [comment]
Okay!

**Comment by tomaarsen** [comment]
The test failures seem unrelated at a glance.

- Tom Aarsen

File: src/transformers/models/eurobert/modular_eurobert.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+            rope_parameters=rope_parameters,
+            attention_bias=attention_bias,
+            attention_dropout=attention_dropout,
+            mlp_bias=mlp_bias,
+            head_dim=head_dim,
+            **kwargs,
+        )
+        self.mask_token_id = mask_token_id
+        self.classifier_pooling = classifier_pooling
+        self.is_causal = False
```
Comment (ArthurZucker) [inline src/transformers/models/eurobert/modular_eurobert.py:169]: not sure this changes anything!

File: tests/models/eurobert/test_modeling_eurobert.py [hunk 3]
```diff
@@ -0,0 +1,591 @@
...
+                base_config, {"rope_scaling": {"rope_type": "linear", "factor": 10.0, "foo": "bar"}}
+            )
+            original_model = EuroBertForMaskedLM(config).to(torch_device)
+            original_model(**model_inputs)
+            self.assertEqual(len(logs.output), 1)
+            self.assertIn("Unrecognized keys", logs.output[0])
+
+        # from a config with specific rope type but missing one of its mandatory parameters -> ❌ throws exception
+        with self.assertRaises(KeyError):
+            config = _reinitialize_config(base_config, {"rope_scaling": {"rope_type": "linear"}})  # missing "factor"
```
Comment (ArthurZucker) [inline tests/models/eurobert/test_modeling_eurobert.py:522]: I don't think you need these TBH!

**Review by ArthurZucker — COMMENTED** [review]
🟢 LGTM

File: src/transformers/models/eurobert/modular_eurobert.py (hunk 2 above)
Comment (tomaarsen) [reply src/transformers/models/eurobert/modular_eurobert.py:169]: It should, due to https://github.com/huggingface/transformers/pull/43705/changes#diff-fca2adc00a38d9fa729672b0d32752b4252a35b1c296397c208ff0f1c764e07bR870 by @Cyrilvallez 's PR to allow bidirectional attention on architectures that are otherwise causal only

File: tests/models/eurobert/test_modeling_eurobert.py (hunk 3 above)
Comment (tomaarsen) [reply tests/models/eurobert/test_modeling_eurobert.py:522]: Agreed, but some of these were already there, and RoPE is mostly tested via the CausalLM tester superclass. Because this model doesn't do CausalLM (and so doesn't use that tester superclass), but does use RoPE in this way, I figured I'd adopt the tests to make sure I'm not doing anything unusual.
